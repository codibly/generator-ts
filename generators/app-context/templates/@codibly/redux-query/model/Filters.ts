export type Filters = {
  [field: string]: any;
};
