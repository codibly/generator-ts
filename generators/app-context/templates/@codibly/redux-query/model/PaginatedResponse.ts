import { Pagination } from './Pagination';

export type PaginatedResponse<T> = {
  data: T[];
  pagination: Pagination.Statistics;
};
