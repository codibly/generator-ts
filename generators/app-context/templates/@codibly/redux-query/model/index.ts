export * from './Filters';
export * from './Pagination';
export * from './PaginatedResponse';
export * from './Sorting';
export * from './Query';
