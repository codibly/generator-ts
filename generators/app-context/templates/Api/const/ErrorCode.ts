
export enum ErrorCode {
  VALIDATION_FAILED = 'VALIDATION_FAILED',
  BAD_REQUEST = 'BAD_REQUEST'
}
