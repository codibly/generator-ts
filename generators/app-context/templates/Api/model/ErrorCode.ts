// we don't have dto for errors code - we will keep it in sync with API
export enum ErrorCode {
  VALIDATION_FAILED = 'VALIDATION_FAILED',
  BAD_REQUEST = 'BAD_REQUEST'
}
