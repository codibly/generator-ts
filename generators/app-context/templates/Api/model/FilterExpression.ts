/**
 * Basic operands
 */
export interface IdentifierOperand {
  identifier: string;
}
export interface LiteralOperand {
  literal: any;
}

/**
 * Basic expressions
 */
export interface IdentifierExpression {
  [0]: IdentifierOperand;
  length: 1;
}
export interface IdentifierLiteralExpression {
  [0]: IdentifierOperand;
  [1]: LiteralOperand;
  length: 2;
}

/**
 * Terminal expressions
 */
export interface NullExpression {
  null: IdentifierExpression;
}
export interface NotNullExpression {
  notNull: IdentifierExpression;
}
export interface LtExpression {
  lt: IdentifierLiteralExpression;
}
export interface LteExpression {
  lte: IdentifierLiteralExpression;
}
export interface GtExpression {
  gt: IdentifierLiteralExpression;
}
export interface GteExpression {
  gte: IdentifierLiteralExpression;
}
export interface InExpression {
  in: IdentifierLiteralExpression;
}
export interface EqExpression {
  eq: IdentifierLiteralExpression;
}
export interface LikeExpression {
  like: IdentifierLiteralExpression;
}
export type TerminalExpression =
  | NullExpression
  | NotNullExpression
  | LtExpression
  | LteExpression
  | GtExpression
  | GteExpression
  | InExpression
  | EqExpression
  | LikeExpression;

/**
 * Logic expressions
 */
export interface AndExpression {
  and: FilterExpression[];
}
export interface OrExpression {
  or: FilterExpression[];
}
export type LogicExpression = AndExpression | OrExpression;

/**
 * Filter expression
 */
export type FilterExpression = LogicExpression | TerminalExpression;

export namespace FilterExpression {
  // type guards
  export function isLikeExpression(filter: FilterExpression): filter is LikeExpression {
    return !!(filter as LikeExpression).like;
  }

  // builders
  export enum LikeMode {
    STARTS_WITH = 'STARTS_WITH',
    CONTAINS = 'CONTAINS',
    EQUALS = 'EQUALS',
    ENDS_WITH = 'ENDS_WITH'
  }

  export function likeLiteral(literal: string, mode: LikeMode): string {
    switch (mode) {
      case LikeMode.STARTS_WITH:
        return `${literal}%`;

      case LikeMode.ENDS_WITH:
        return `%${literal}`;

      case LikeMode.CONTAINS:
        return `%${literal}%`;

      case LikeMode.EQUALS:
      default:
        return literal;
    }
  }

  // we can't use "in" as it's keyword in the JavaScript
  export function oneOf(identifier: string, literal: any[]): InExpression {
    return {
      in: [{ identifier }, { literal }]
    };
  }

  export function eq(identifier: string, literal: any): EqExpression {
    return {
      eq: [{ identifier }, { literal }]
    };
  }

  export function like(
    identifier: string,
    query: string,
    mode: LikeMode = LikeMode.CONTAINS
  ): LikeExpression {
    return {
      like: [{ identifier }, { literal: likeLiteral(query, mode) }]
    };
  }

  export function or(filters: FilterExpression[]): FilterExpression {
    if (filters.length === 1) {
      return filters[0];
    } else {
      return {
        or: filters
      };
    }
  }

  export function and(filters: FilterExpression[]): FilterExpression {
    if (filters.length === 1) {
      return filters[0];
    } else {
      return {
        and: filters
      };
    }
  }
}
