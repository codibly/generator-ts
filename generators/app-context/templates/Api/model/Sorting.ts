import { ApiDto } from '../api/Api.dto';
import { SortingDto } from '../api/Sorting/Sorting.dto';

// we keep model and DTO in sync
export type Sorting = SortingDto;

export namespace Sorting {
  export type Direction = SortingDto.Direction;

  export function flipDirection(direction: Direction | undefined): Direction {
    return direction === 'asc' ? 'desc' : 'asc';
  }

  /**
   * Flip existing sorting direction
   */
  export function flip<T extends Sorting | null>(sorting: T): T {
    return sorting
      ? ({ field: sorting.field, direction: flipDirection(sorting.direction) } as T)
      : (null as T);
  }

  /**
   * Create sorting object based on field and direction (by default asc)
   */
  export function create(field: string, direction: Direction = 'asc'): Sorting {
    return { field, direction };
  }
}
