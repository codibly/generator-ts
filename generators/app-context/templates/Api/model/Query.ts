import { FilterExpression } from './FilterExpression';
import { Pagination } from './Pagination';
import { Sorting } from './Sorting';

export type Query = {
  filter?: FilterExpression | null;
  pagination?: Pagination;
  sorting?: Sorting | null;
};
