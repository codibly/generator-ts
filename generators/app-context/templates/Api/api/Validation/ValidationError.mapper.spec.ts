import { ErrorCode } from '../../model/ErrorCode';
import { ValidationErrorDto } from './ValidationError.dto';
import { ValidationErrorMapper } from './ValidationError.mapper';

// tslint:disable:no-duplicate-string
describe('ValidationErrorMapper', () => {
  it('should map nested errors', () => {
    const error: ValidationErrorDto = {
      errorCode: 'VALIDATION_FAILED' as ErrorCode.VALIDATION_FAILED,
      message: 'Validation failed',
      violations: [
        {
          message: 'This value \u0022suite\u0022 should not be null ',
          property: 'address.suite',
          meta: {
            template: 'This value {{ value }} should not be null ',
            params: {
              '{{ value }}': '\u0022suite\u0022'
            }
          }
        }
      ]
    };

    expect(ValidationErrorMapper.toFormErrors(error)).toMatchSnapshot();
  });

  it('should map nested sheet errors', () => {
    const error: ValidationErrorDto = {
      errorCode: 'VALIDATION_FAILED' as ErrorCode.VALIDATION_FAILED,
      message: 'Validation failed',
      violations: [
        {
          message: 'This value \u0022name\u0022 should not be null ',
          property: 'name',
          meta: {
            template: 'This value {{ value }} should not be null ',
            params: {
              '{{ value }}': '\u0022name\u0022'
            }
          }
        },
        {
          message: 'This value should be 0 or more.',
          property: 'sheet.D2',
          meta: {
            template: 'This value should be {{ limit }} or more.',
            params: {
              '{{ value }}': '-5000000',
              '{{ limit }}': '0'
            }
          }
        }
      ]
    };

    expect(ValidationErrorMapper.toFormErrorsWithSheetCells(error, ['sheet'])).toMatchSnapshot();
  });
});
