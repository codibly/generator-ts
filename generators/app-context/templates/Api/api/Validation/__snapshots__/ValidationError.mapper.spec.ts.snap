// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValidationErrorMapper should map nested errors 1`] = `
Object {
  "address": Object {
    "suite": "This value \\"suite\\" should not be null ",
  },
}
`;

exports[`ValidationErrorMapper should map nested sheet errors 1`] = `
Object {
  "name": "This value \\"name\\" should not be null ",
  "sheet": "Cell D2: This value should be 0 or more.",
}
`;
